src/Yarhl/IO/DataStream.cs

```diff
@@ -552,11 +568,15 @@ protected virtual void Dispose(bool freeManagedResourcesAlso)
 
             Disposed = true;
 
-            if (BaseStream != null && hasOwnsership) {
-                Instances[BaseStream] -= 1;
+            if (BaseStream == null || !hasOwnsership)
+                return;
+
+            Instances[BaseStream] -= 1;
```

Comment:
I think we need to put this one inside the `lock` in case the thread inside the `lock` has already remove it.

Reply:
I think it doesn't matter if it is outside the lock. Inside the lock, it checks if `Instances[BaseStream] == 0` before Dispose, so there can't be other threads trying to dispose the BaseStream.